README

I sent ohlc.pm to Benjamin Warfield (cpan/~bwarfield) about a
year ago.  He later told me that he gets millions of patches,
some that work, some that don't and that he doesn't really have
time to apply them all.  This is understandable.  He also told me
that making sure GD::Graph is stable on all 16,000 platforms is
his number one goal.  Admirable.

He then pointed out that I didn't really need to wait on him... I
could just release a sub-module.  I don't know why that didn't
occur to me in the first place.  Doi.

Anyway, what is an OHLC graph?  Let the example speak for itself:

    https://voltar.org/PerlModules/ohlc_example.png

    (This is a mixed.pm graph.  The lines with ticks are the
    ohlc chart.  The ticks show the opening and closing
    prices of MSFT.  The line itself shows the high and
    low for the period.)



Changes
0.9607: Sun Jan 06 2013
   - major bugs relating to the alternate color stuff that I
     don't think anyone (including myself) ever used.  Stripped
     it out.
   - Zen Kurokawa (@gmail) points out major drawing failure when
     the candlesticks when there are fewer data points than
     needed to fill the graph.

     https://rt.cpan.org/Ticket/Display.html?id=82524

     my fix (temporary?) disables the {overwrite} functionality
     introduced by andrew+github@afresh1.com

0.9606: Mon Oct 10 2011
   - contribs for candlestick overwrite mode, thanks Andrew

0.9602: Fri Oct 09 2009
   - candlesticks were upside down -- how did I not notice that?
     Thanks Nilson.

0.9602: Thu Aug 27 2009
   - Made a few pod changes (in addition to Ken's) and moved a
     few files around in the dist.

0.9602: Fri Aug 21 2009
   - Ken Tuchman <tuchman at t-online tld=de> pointed out some
     flaws in the PODs.

0.9601: Wed Jul 08 2009
   - Doc bugs (very indirectly) found by Hans Dieter Pearcey
   - added critic.t and fixed many things

0.9600: Sun Apr 26 17:56:40 EDT 2009
   - started adding candlesticks
   - added the pod and the pod-example test script
   - got candlesticks working -- seems to work fine in mixed mode too

0.9402: Wed Apr 15 22:05:18 EDT 2009
   - misnamed in the name line of the pod.  Wow.

0.9401: Wed Apr 15 07:29:37 EDT 2009
   - this seems more or less ready to distribute

0.9401: Wed Apr 15 06:53:15 EDT 2009
   - got this more or less ready to try to use

0.01: Wed Apr 15 05:46:18 EDT 2009
   - skelling this up


ohlc.pm
#==========================================================================
#              Copyright (c) 2008 Paul Miller
#==========================================================================

package GD::Graph::ohlc;

use strict;
use warnings;

use GD::Graph::mixed; # NOTE: we pull this in so we can modify part of it.
use GD::Graph::axestype;
use GD::Graph::utils qw(:all);
use GD::Graph::colour qw(:colours);

use constant PI => 4 * atan2(1,1);

our $VERSION = "0.9607";
our @ISA = qw(GD::Graph::axestype);

push @GD::Graph::mixed::ISA, __PACKAGE__;

# draw_data_set {{{
sub draw_data_set {
    my $self = shift;
    my $ds   = shift;

    my @values = $self->{_data}->y_values($ds) or
        return $self->_set_error("Impossible illegal data set: $ds", $self->{_data}->error);

    # Pick a colour
    my $dsci = $self->set_clr($self->pick_data_clr($ds));

    my $GX;
    my ($ox,$oy, $cx,$cy, $lx,$ly, $hx,$hy); # NOTE: all the x's are the same...
    for (my $i = 0; $i < @values; $i++) {
        my $value = $values[$i];
        next unless ref($value) eq "ARRAY" and @$value==4;
        my ($open, $high, $low, $close) = @$value;

        if (defined($self->{x_min_value}) && defined($self->{x_max_value})) {
            $GX = $self->{_data}->get_x($i);

            ($ox, $oy) = $self->val_to_pixel($GX, $value->[0], $ds);
            ($hx, $hy) = $self->val_to_pixel($GX, $value->[1], $ds);
            ($lx, $ly) = $self->val_to_pixel($GX, $value->[2], $ds);
            ($cx, $cy) = $self->val_to_pixel($GX, $value->[3], $ds);

        } else {
            ($ox, $oy) = $self->val_to_pixel($i+1, $value->[0], $ds);
            ($hx, $hy) = $self->val_to_pixel($i+1, $value->[1], $ds);
            ($lx, $ly) = $self->val_to_pixel($i+1, $value->[2], $ds);
            ($cx, $cy) = $self->val_to_pixel($i+1, $value->[3], $ds);
        }

        $self->ohlc_marker($ox,$oy, $cx,$cy, $lx,$ly, $hx,$hy, $dsci );
        $self->{_hotspots}[$ds][$i] = ['rect', $self->ohlc_marker_coordinates($ox,$oy, $cx,$cy, $lx,$ly, $hx,$hy)];
    }

    return $ds;
}
# }}}
# ohlc_marker_coordinates {{{
sub ohlc_marker_coordinates {
    my $self = shift;
    my ($ox,$oy, $cx,$cy, $lx,$ly, $hx,$hy) = @_;

    return ( $ox-2, $cx+2, $hy, $ly );
}
# }}}
# ohlc_marker {{{
sub ohlc_marker {
    my $self = shift;
    my ($ox,$oy, $cx,$cy, $lx,$ly, $hx,$hy, $mclr) = @_;
    return unless defined $mclr;

    $self->{graph}->line( ($ox,$oy) => ($ox-2,$oy), $mclr );
    $self->{graph}->line( ($cx,$cy) => ($cx+2,$cy), $mclr );
    $self->{graph}->line( ($lx,$ly) => ($hx,$hy),   $mclr );

    return;
}
# }}}

1;


candlesticks.pm
#==========================================================================
#              Copyright (c) 2008 Paul Miller
#==========================================================================

package GD::Graph::candlesticks;

use strict;
use warnings;

use GD::Graph::mixed; # NOTE: we pull this in so we can modify part of it.
use GD::Graph::axestype;
use GD::Graph::utils qw(:all);
use GD::Graph::colour qw(:colours);

use constant PI => 4 * atan2(1,1);

our $VERSION = "0.9607";
our @ISA = qw(GD::Graph::axestype);

our %DEFAULT = (
    correct_width => 1,
    candlestick_width => 7,
    candlestick_spacing     => 0,
    candlestickgroup_spacing=> 25,
);

push @GD::Graph::mixed::ISA, __PACKAGE__;

# working off gdgraph/Graph/bars.pm (in addition to ohlc.pm)

# initialise {{{
sub initialise {
    my $self = shift;

    $self->SUPER::initialise();

    while (my($key, $val) = each %DEFAULT)
        { $self->{$key} = $val }

    return 1;
}
# }}}
# _has_default {{{
sub _has_default {
    my $this = shift;

    return $DEFAULT{$_[0]} if exists $DEFAULT{$_[0]};
    return $this->SUPER::_has_default(@_);
}
# }}}
# draw_data_set {{{
sub draw_data_set {
    my $this = shift;
    my $ds   = shift;

    my @values = $this->{_data}->y_values($ds) or
        return $this->_set_error("Impossible illegal data set: $ds", $this->{_data}->error);

    # Pick a colour
    my $dsci = $this->set_clr($this->pick_data_clr($ds));

    my $GX;
    my ($ox,$oy, $cx,$cy, $lx,$ly, $hx,$hy); # NOTE: all the x's are the same...
    for (my $i = 0; $i < @values; $i++) {
        my $value = $values[$i];
        next unless ref($value) eq "ARRAY" and @$value==4;
        my ($open, $high, $low, $close) = @$value;

        if (defined($this->{x_min_value}) && defined($this->{x_max_value})) {
            $GX = $this->{_data}->get_x($i);

            ($ox, $oy) = $this->val_to_pixel($GX, $value->[0], $ds);
            ($hx, $hy) = $this->val_to_pixel($GX, $value->[1], $ds);
            ($lx, $ly) = $this->val_to_pixel($GX, $value->[2], $ds);
            ($cx, $cy) = $this->val_to_pixel($GX, $value->[3], $ds);

        } else {
            ($ox, $oy) = $this->val_to_pixel($i+1, $value->[0], $ds);
            ($hx, $hy) = $this->val_to_pixel($i+1, $value->[1], $ds);
            ($lx, $ly) = $this->val_to_pixel($i+1, $value->[2], $ds);
            ($cx, $cy) = $this->val_to_pixel($i+1, $value->[3], $ds);
        }

     #  if (!$this->{overwrite}) {
     #      my $candlestick_s = $this->{candlestick_spacing}/2;
     #      my $window = $this->{x_step} - $this->{candlestickgroup_spacing};
     #
     #      foreach my $x ($ox, $hx, $lx, $cx) {
     #          $x = $x
     #          - $window/2
     #          + ($ds - 1) * $window/$this->{_data}->num_sets
     #          + $candlestick_s + 1;
     #      }
     #  }

        $this->candlesticks_marker($ox,$oy, $cx,$cy, $lx,$ly, $hx,$hy, $dsci );
        $this->{_hotspots}[$ds][$i] = ['rect', $this->candlesticks_marker_coordinates($ox,$oy, $cx,$cy, $lx,$ly, $hx,$hy)];
    }

    return $ds;
}
# }}}
# half_width {{{
sub half_width {
    my $this = shift;

    return int( $this->{candlestick_width} / 2 ) if exists $this->{candlestick_width};
    return 3;
}
# }}}
# candlesticks_marker_coordinates {{{
sub candlesticks_marker_coordinates {
    my $this = shift;
    my ($ox,$oy, $cx,$cy, $lx,$ly, $hx,$hy) = @_;

    my $h = $this->half_width;
    return ( $ox - $h, $cx + $h, $hy, $ly );
}
# }}}
# candlesticks_marker {{{
sub candlesticks_marker {
    my $this = shift;
    my ($ox,$oy, $cx,$cy, $lx,$ly, $hx,$hy, $mclr) = @_;
    return unless defined $mclr;

    $this->{graph}->line( ($lx,$ly) => ($hx,$hy), $mclr );

    my $h = $this->half_width;
    if( $cy>$oy ) {
        $this->{graph}->filledRectangle( ($cx - $h, $cy) => ($ox + $h, $oy), $mclr );

    } else {
        $this->{graph}->filledRectangle( ($cx - $h, $cy) => ($ox + $h, $oy), $this->{bgci} );
        $this->{graph}->rectangle(       ($cx - $h, $cy) => ($ox + $h, $oy), $mclr );
    }

    return;
}
# }}}

1;


mixed_candle.pl
#!/usr/bin/perl -Iblib/lib

use IPC::System::Simple qw(systemx);

BEGIN {
    my @mtime = map {(stat $_)[9]} qw(Makefile Makefile.PL);
    systemx(qw(perl Makefile.PL)) if $mtime[0] != $mtime[1];
    systemx(qw(make -f Makefile));
}

use strict;
use warnings;
use List::Util qw(min max);
use GD::Graph::candlesticks;
use GD::Graph::mixed;

my @msft = (     #  open       high       low        close
    ["2007/12/18", "34.6400", "35.0000", "34.2100", "34.7400"],
    ["2007/12/19", "34.6900", "35.1400", "34.3800", "34.7900"],
    ["2007/12/20", "35.2900", "35.7900", "35.0800", "35.5200"],
    ["2007/12/21", "35.9000", "36.0600", "35.7500", "36.0600"],
    ["2007/12/24", "36.1300", "36.7200", "36.0500", "36.5800"],
    ["2007/12/26", "36.4100", "36.6400", "36.2600", "36.6100"],
    ["2007/12/27", "36.3500", "36.5500", "35.9400", "35.9700"],
    ["2007/12/28", "36.1000", "36.2300", "35.6700", "36.1200"],
    ["2007/12/31", "35.9000", "35.9900", "35.5200", "35.6000"],
    ["2008/01/02", "35.7900", "35.9600", "35.0000", "35.2200"],
    ["2008/01/03", "35.2200", "35.6500", "34.8600", "35.3700"],
    ["2008/01/04", "35.1900", "35.2000", "34.0900", "34.3800"],
    ["2008/01/07", "34.5500", "34.8000", "34.2500", "34.6100"],
    ["2008/01/08", "34.7100", "34.7100", "33.4000", "33.4500"],
    ["2008/01/09", "33.3600", "34.5400", "33.3500", "34.4400"],
    ["2008/01/10", "34.3500", "34.5000", "33.7800", "34.3300"],
    ["2008/01/11", "34.1400", "34.2400", "33.7200", "33.9100"],
    ["2008/01/14", "34.4600", "34.5700", "34.0800", "34.3900"],
    ["2008/01/15", "34.0300", "34.3800", "34.0000", "34.0000"],
    ["2008/01/16", "33.4200", "33.6500", "32.5100", "33.2300"],
    ["2008/01/17", "33.5400", "33.8000", "32.9700", "33.1100"],
    ["2008/01/18", "33.1600", "34.0000", "32.9700", "33.0100"],
    ["2008/01/22", "31.5400", "32.5300", "31.5000", "31.9600"],
    ["2008/01/23", "31.4800", "32.0500", "31.0400", "31.9300"],
    ["2008/01/24", "32.3500", "33.3600", "32.1200", "33.2500"],
    ["2008/01/25", "34.9000", "35.0000", "32.8700", "32.9400"],
    ["2008/01/28", "33.0200", "33.1000", "32.4200", "32.7200"],
    ["2008/01/29", "32.8500", "32.8900", "32.3500", "32.6000"],
    ["2008/01/30", "32.5600", "32.8000", "32.0500", "32.2000"],
    ["2008/01/31", "31.9100", "32.7400", "31.7200", "32.6000"],
);

my @all_points = map {@$_[1 .. 4]} @msft[1 .. $#msft];
my $min_point  = min(@all_points);
my $max_point  = max(@all_points);

my $graph = GD::Graph::mixed->new(800, 400);
   $graph->set( 
        x_labels_vertical => 1,
        x_label           => 'Trade Date',
        y_label           => 'NASDAQ:MSFT',
        title             => "Example OHLC",
        transparent       => 0,
        dclrs             => [qw(lgray blue)],
        types             => [qw(lines candlesticks)],
        y_min_value       => $min_point-0.2,
        y_max_value       => $max_point+0.2,
        y_number_format   => '%0.2f',

    ) or warn $graph->error;

my $data_candlesticks = [
    [ map {$_->[0]} @msft ],       # date
    [ map {$_->[4]} @msft ],       # close
    [ map {[@$_[1 .. 4]]} @msft ], # candlesticks
];

my $gd = $graph->plot($data_candlesticks) or die $graph->error;
open my $dump, ">candlemixed_example.png" or die $!;
print $dump $gd->png;
close $dump;
